src/Access/DiskAccessStorage.cpp

```diff
@@ -386,21 +379,85 @@ bool DiskAccessStorage::rebuildLists()
         if (!entity)
             continue;
 
-        const String & name = entity->getName();
-        auto type = entity->getType();
-        auto & entry = entries_by_id[id];
-        entry.id = id;
-        entry.type = type;
-        entry.name = name;
-        entry.entity = entity;
-        auto & entries_by_name = entries_by_name_and_type[static_cast<size_t>(type)];
-        entries_by_name[entry.name] = &entry;
+        all_entities.emplace_back(id, entity);
     }
 
+    setAll(all_entities);
+
     for (auto type : collections::range(AccessEntityType::MAX))
         types_of_lists_to_write.insert(type);
 
-    return true;
+    failed_to_write_lists = false; /// Try again writing lists.
+    writeLists();
+}
+
+void DiskAccessStorage::setAll(const std::vector<std::pair<UUID, AccessEntityPtr>> & all_entities)
+{
+    /// This function is similar to MemoryAccessStorage::setAll().
+    boost::container::flat_set<UUID> not_used_ids;
+    std::vector<UUID> conflicting_ids;
+
+    /// Get the list of currently used IDs. Later we will remove those of them which are not used anymore.
+    for (const auto & [id, _] : entries_by_id)
+        not_used_ids.emplace(id);
+
+    /// Get the list of conflicting IDs and update the list of currently used ones.
+    for (const auto & [id, entity] : all_entities)
+    {
+        auto it = entries_by_id.find(id);
+        if (it != entries_by_id.end())
+        {
+            not_used_ids.erase(id); /// ID is used.
+
+            const Entry & entry = it->second;
+            if (entry.entity->getType() != entity->getType())
+                conflicting_ids.emplace_back(id); /// Conflict: same ID, different type.
+        }
+
+        const auto & entries_by_name = entries_by_name_and_type[static_cast<size_t>(entity->getType())];
+        auto it2 = entries_by_name.find(entity->getName());
+        if (it2 != entries_by_name.end())
+        {
+            const Entry & entry = *(it2->second);
+            if (entry.id != id)
+                conflicting_ids.emplace_back(entry.id); /// Conflict: same name and type, different ID.
+        }
+    }
+
+    /// Remove entities which are not used anymore and which are in conflict with new entities.
+    boost::container::flat_set<UUID> ids_to_remove = std::move(not_used_ids);
+    boost::range::copy(conflicting_ids, std::inserter(ids_to_remove, ids_to_remove.end()));
+    for (const auto & id : ids_to_remove)
+        removeNoLock(id, /* throw_if_not_exists= */ false);
+
+    /// Insert or update entities.
+    for (const auto & [id, entity] : all_entities)
+    {
+        auto it = entries_by_id.find(id);
+        if (it == entries_by_id.end())
```

Comment:
```suggestion
        if (auto it = entries_by_id.find(id);
            it == entries_by_id.end())
```
I think also should work here